UOP1_Project/Assets/Scripts/Sound/SoundManager.cs

```diff
@@ -0,0 +1,190 @@
+using UnityEngine;
+using UnityEngine.Audio;
+
+public class SoundManager : MonoBehaviour
+{
+	public static SoundManager singleton;
+
+	[Header ("Init Variables")]
+	[Tooltip ("Amount of sound emitters created on Start")]
+	public int initialPoolSize = 1;
+
+	[Header ("Mixer Groups")]
+	public AudioMixerGroup masterMixer;
+	public AudioMixerGroup sfxMixer;
+	public AudioMixerGroup musicMixer;
+
+	private List<SoundEmitter> soundEmitterPool = new List<SoundEmitter>();
+	private GameObject soundEmittersContainer;
+	private int poolIndex = 0;
+
+	private void Awake()
+	{
+		DontDestroyOnLoad(this);
+
+		if (!SoundManager.singleton)
+			singleton = this;
+		else
+		{
+			Debug.LogWarning("One sound manager already present in the game, destroying " + this.name + ".");
+			Destroy(this);
+		}
+	}
+
+	void Start()
+    {
+		InitEmitterPool();
+    }
+
+	#region mixer groups functions
+	public static void ChangeVolumeOfMixerGroup(AudioMixerGroup mixer, float newVolumeNormalized)
+	{
+		mixer.audioMixer.SetFloat("Volume", NormalizedToMixerValue(newVolumeNormalized));
+	}
+
+	public static void SaveVolumeOfMixerGroup(AudioMixerGroup mixer)
+	{
+		float tempVolume;
+
+		if (mixer.audioMixer.GetFloat("Volume", out tempVolume))
+		{
+			PlayerPrefs.SetFloat(mixer.name, tempVolume);
+		}
+		else
+			Debug.LogError("Could not save volume for mixer group " + mixer.name + ". It does not contain an exposer variable with name Volume" );
+	}
+
+	public static void LoadVolumeOfMixerGroup(AudioMixerGroup mixer)
+	{
+		float tempVolume = PlayerPrefs.GetFloat(mixer.name, Mathf.Infinity);
+
+		if (tempVolume != Mathf.Infinity)
+		{
+			mixer.audioMixer.SetFloat("Volume", tempVolume);
+		}
+		else
+			Debug.Log("There is no saved volume preferences for mixer " + mixer.name + ", could not load volume.");
+	}
+
+	#region mixerHelpers
+	// Both MixerValueNormalized and NormalizedToMixerValue functions are used for easier transformations when using UI sliders normalized format
+	private static float MixerValueNormalized(float value)
+	{
+		return  (-(value - 80) / 80) - 1;
+	}
+	private static float NormalizedToMixerValue(float normalizedValue)
+	{
+		return -80 + (normalizedValue * 80);
+	}
+	#endregion
+	#endregion
+
+	#region play sounds functions
+	public SoundEmitter Play2DSound(AudioClip clip, AudioMixerGroup mixer, bool loop = false, float volume = 1, float pitch = 1)
+	{
+		if (!mixer)
+			mixer = masterMixer;
+
+		SoundEmitter foundSoundEmit = GetSoundEmitter();
+
+		foundSoundEmit.PlaySound(clip, mixer, Vector3.zero, loop, volume, pitch, 0);
+
+		return foundSoundEmit;
+	}
+
+	public SoundEmitter PlaySpatialSound(AudioClip clip, AudioMixerGroup mixer, Vector3 position, bool loop = false, float volume = 1, float pitch = 1, float spartialBlend = 1)
+	{
+		if (!mixer)
+			mixer = masterMixer;
+
+		SoundEmitter foundSoundEmit = GetSoundEmitter();
+
+		foundSoundEmit.PlaySound(clip, mixer, position, loop, volume, pitch, spartialBlend);
+
+		return foundSoundEmit;
+	}
```

Comment:
Well, i thought about doing it that way as well and that is a possibility. Nonetheless does it make sense to have two methods here to split into 2D and 3D for clarity of purpose of those methods. These methods are public API so to speak after all. Other people will be using them and the usage should be as streamlined as possible.

I would prefer leaving this as is.

Reply:
I see what you are saying, but consider this - we may eventually want to be able to specify the values of many more AudioSource settings. I'd say there are about 19 relevant settings that one could conceivably want to change. Instead of writing a method signature 21 params long, it would make more sense to wrap those in an AudioSourceSettings structure with Unity's own defaults, let the user override the ones they want, and the pass the whole structure to be applied.

The method signature would then look like this: 
`public void PlaySound(AudioClip clip, AudioSourceSettings settings, Vector3 position = default)`

Usage for 2D sounds:
```
_soundEmitter.PlaySound(clip,AudioSourceSettings.Default);
```
The user just passes in the clip and default settings (because the default is no spatialization). Position has the default of Vector.zero and does not need to be specified in this case.

Usage for 3D sounds:
```
_soundEmitter.PlaySound(clip,new AudioSourceSettings{SpatialBlend=1f},transform.position);
```
The user overrides the spatial blend and sets a position.

Possibly more importantly, this would match Unity's own API. They do not have separate Play methods for 2D vs 3D. The spatialization is done via fields on the AudioSource component.

Reply:
This was something that had to be done eventually for the reasons you stated. I thought of doing a struct as the settings as a start but decided against it to maintain a more simple PR to start.

Your implementation looks really good, if you are working on this on your repository I'll get to work tomorrow with fleshing out the pooler with what we discussed before.